### PR TITLE
SAM predictor & auto annotation

### PR DESCRIPTION
**Pull request by AyushExel**
@Laughing-q opening this up
New Sam Models for Inference
1. from ultralytics import SAM
```
img = cv2.imread("ultralytics/assets/zidane.jpg")
model = SAM("sam_b.pt")
results = model.predict("ultralytics/assets/zidane.jpg", device='cpu')
plot_img = results[0].plot()
```

2. Generate Segmentation dataset labels from images (Auto annotate)
 ```
from ultralytics.yolo.data.annotator import auto_annotate

auto_annotate(data="ultralytics/assets", output_dir="../labels")
```

### Summary
🎨🧠🖼️

This pull request adds a new feature to the ultralytics repository, which is the SAM model for segmenting anything in an image using a vision transformer backbone. The pull request adds several new files and classes to the `ultralytics/vit/sam` subpackage, which implement the SAM model architecture, the automatic mask generator, the predictor, and the utilities for building, transforming, and processing the inputs and outputs. The pull request also modifies some existing files in the `ultralytics/vit` and `ultralytics/yolo` subpackages, to enable importing and using the SAM model and predictor, and to handle the case of mask-only results.

> _`SAM` segments images_
> _With prompts and transformers_
> _Autumn leaves of masks_

### Walkthrough
*  Add SAM model and predictor for segmenting anything in an image using a vision transformer backbone ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-7bb55fea226aca9e349baf67b2a6011df4f4a9b5863393d60666c22575cef95eR498-R532), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-5597b74444fe27a26922336224a04bd9ed28414e3b2d8eb28dc09f68cf9e367dR1), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-d5cafb522151f584a4dcef1ddbdc85c3533ce91140f5fa4a37153858f9ae4898R1), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-74c47ae62a03b5d98f69b18664373db3ff634fa8fd17b5f721c91b123ec6abf6R1-R43), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-8b1a338736ea0a3934448b5ce4cdb55b827daa428faac0b6b909cb4b185a8232R1-R16), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-0b7382860c30b3bf3b9d77026476060bebde600926c8a3067eac315911abe51bR1-R319), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-1eca6cd776e7349f231543ef47fd51a0711ecd87dcc29ce6c50ccb4befd03dafR1-R92), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-d7fc935999f7e83894d6df9af3a146e4af7358c71ac0cbd807499ddecb48ee63R1-R114), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-2d39b5e0456d1c175512511d9275edff5ee561dfad71eee392f031be115f47f4R1-R4), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-864a45978308bbb8e20228be1616336c9f7bf957aae4329cb3b681ab967d012cR1-R164), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-8ce06827c5a9984fd70865dfdaf2df8e0634dfd10687c9c6a46321435b013043R1-R351), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-6012fa7d59126856cee037e985d006d7e6192cdd6dec4ef30d23aba7c3e96164R1-R170), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-7c09922ca0cb3164343a7a8caf0d5d19b7ed45938e33a4154bb2fedc38e38423R1-R256), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-88e9b2ce45939d07a4f76051710127e57ab48c3c713f5039e627f2878ad5865fR1-R227), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-6e94155666f26dda776758e4ea267377fd4c50f6f5c62f7b56e7886105dcc1c8R1-R31))
  * Define the `Sam` class as the main module for the SAM model, which takes an image encoder, a prompt encoder, and a mask decoder as arguments ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-6012fa7d59126856cee037e985d006d7e6192cdd6dec4ef30d23aba7c3e96164R1-R170))
  * Implement the image encoder, prompt encoder, and mask decoder modules using different vision transformer backbones and a two-way transformer architecture ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-864a45978308bbb8e20228be1616336c9f7bf957aae4329cb3b681ab967d012cR1-R164), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-88e9b2ce45939d07a4f76051710127e57ab48c3c713f5039e627f2878ad5865fR1-R227))
  * Provide functions for building SAM models from pre-trained checkpoints with different backbones ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-d7fc935999f7e83894d6df9af3a146e4af7358c71ac0cbd807499ddecb48ee63R1-R114))
  * Define the `SamPredictor` class as a module that allows efficient mask prediction given an image and prompts ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-7c09922ca0cb3164343a7a8caf0d5d19b7ed45938e33a4154bb2fedc38e38423R1-R256))
  * Define the `SamAutomaticMaskGenerator` class as a module that generates masks for the entire image using a SAM model ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-8ce06827c5a9984fd70865dfdaf2df8e0634dfd10687c9c6a46321435b013043R1-R351), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-0b7382860c30b3bf3b9d77026476060bebde600926c8a3067eac315911abe51bR1-R319))
  * Define the `Predictor` class as a subclass of `BasePredictor` that uses a SAM model to generate masks for an image, given a source path or array ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-74c47ae62a03b5d98f69b18664373db3ff634fa8fd17b5f721c91b123ec6abf6R1-R43))
  * Define the `ResizeLongestSide` class as a transformation that resizes images to the longest side target length ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-1eca6cd776e7349f231543ef47fd51a0711ecd87dcc29ce6c50ccb4befd03dafR1-R92))
  * Define the `sam` function as a convenience wrapper for building a SAM model from a checkpoint name ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-8b1a338736ea0a3934448b5ce4cdb55b827daa428faac0b6b909cb4b185a8232R1-R16))
  * Import the relevant modules and classes from the `ultralytics/vit/sam` subpackage to the `ultralytics/vit` namespace ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-5597b74444fe27a26922336224a04bd9ed28414e3b2d8eb28dc09f68cf9e367dR1), [link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-2d39b5e0456d1c175512511d9275edff5ee561dfad71eee392f031be115f47f4R1-R4))
  * Import the `Predictor` class from the `ultralytics/vit/engine/predictor.py` file to the `ultralytics/vit/engine` namespace ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-d5cafb522151f584a4dcef1ddbdc85c3533ce91140f5fa4a37153858f9ae4898R1))
* Modify the `plot` function in the `Results` class to handle the case when there are no predicted boxes but there are predicted masks ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-c3050afb0dfa35a0115c36c0acd91e2587f4349023d5f8a47e8c44e6b30121c5L216-R217))
  * Format the output of the SAM predictor as a `Results` object, which can be plotted or saved ([link](https://github.com/ultralytics/ultralytics/pull/2232/files?diff=unified&w=0#diff-6e94155666f26dda776758e4ea267377fd4c50f6f5c62f7b56e7886105dcc1c8R1-R31))

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
This PR introduces enhancements and new features to Ultralytics repositories, focusing on testing improvements, SAM (Segment Any Model) integration, MLP and LayerNorm additions, and dataset utilities.

### 📊 Key Changes
- Improved test coverage with additional assertions and fixes.
- Integrated SAM (Segment Anything Model) into the Ultralytics ecosystem.
- Added MLP (multi-layer perceptron) and LayerNorm2d modules for neural networks.
- Enhanced dataset utility scripts for annotation and data loading.
- Updated dependencies and codebase to support new features and improvements.

### 🎯 Purpose & Impact
- **Improved Testing**: Ensures code reliability and reduces the likelihood of bugs reaching production.
- **SAM Integration**: Provides advanced segmentation capabilities, expanding the range of tasks Ultralytics models can perform.
- **New Modules**: MLP and LayerNorm2d support the development of more versatile and effective neural network architectures.
- **Dataset Utilities**: Streamlines data preparation, enabling users to focus on model training and experimentation without worrying about data formatting.

These changes enhance the Ultralytics AI capabilities, making it a more robust and versatile tool for machine learning practitioners.

### CODE REVIEW

**Comment by Laughing-q** [comment]
usage example(temporarily):
```python
import cv2
from ultralytics.vit import SAM

img = cv2.imread("ultralytics/assets/zidane.jpg")
model = SAM("sam_b.pt")
results = model.predict("ultralytics/assets/zidane.jpg", device=0)
plot_img = results.plot()
cv2.imshow('p', plot_img)
cv2.waitKey(0)
```

**Comment by glenn-jocher** [comment]
@AyushExel @Laughing-q awesome! I'll clean up the pre-commit import errors.

**Comment by AyushExel** [comment]
@glenn-jocher yes using it on 1080px images on cpu won't make sense. But it's just a start towards ultralytics Vision transformers :)

**Comment by Laughing-q** [comment]
@AyushExel @glenn-jocher now SAM supports multi-source, like video/image folder, but it's very slow in my test. Also supports out internal args like `save`, `show`.
```python
from ultralytics import SAM

model = SAM("sam_b.pt")
results = model.predict(source=..., device=0, save=True)
```
and also please notice the usage of result.plot() also becomes the same as other tasks:
```python
from ultralytics import SAM

model = SAM("sam_b.pt")
results = model.predict("ultralytics/assets/zidane.jpg", device=0)
plot_img = results[0].plot()   # the above example usage is results.plot()
cv2.imshow('p', plot_img)
cv2.waitKey(0)
```
Despite the usage looks good, there still are many TODOs that I hardcoded some parts of sam predictor. We still need to clean the code.

**Comment by Laughing-q** [comment]
@glenn-jocher CI failed at some pip issues?🤔
image

**Comment by AyushExel** [comment]
@Laughing-q thanks! I'm heading out now. I'll take a look at this tonight

**Comment by glenn-jocher** [comment]
> @glenn-jocher CI failed at some pip issues?🤔 image

Strange. I'm re-running.

**Comment by glenn-jocher** [comment]
@Laughing-q this is awesome. Yes much better to standardize the output into a list for uniformity and consistency.

**Comment by glenn-jocher** [comment]
That's strange, the broken link checker picked up Meituan site as 404, and it's correct:
https://about.meituan.com/en

I'll change it to this site, seems to be working. https://about.meituan.com/

**Comment by Laughing-q** [comment]
@glenn-jocher CI should be fixed now. I'll take a look again tomorrow if it's not😂.

**Comment by glenn-jocher** [comment]
@Laughing-q it's working! Nice work.

**Comment by Laughing-q** [comment]
I found a silence bug yesterday that we don't point the task in Predictor, we point task in model.predict() first then create corresponding Predictor so usually users do not encounter any issues.
But when we create `ClassificationPredictor` for inference, its task actually is the default value `detect`.
Fixed in [https://github.com/ultralytics/ultralytics/pull/2232/commits/33b93272dcb879a4712999ea43515a1915e4e83b](https://github.com/ultralytics/ultralytics/pull/2232/commits/33b93272dcb879a4712999ea43515a1915e4e83b)

**Comment by AyushExel** [comment]
1. New Sam Models for Inference

```
from ultralytics import SAM

img = cv2.imread("ultralytics/assets/zidane.jpg")
model = SAM("sam_b.pt")
results = model.predict("ultralytics/assets/zidane.jpg", device='cpu')
plot_img = results[0].plot()
```

2. Generate Segmentation dataset labels from images (Auto annotate)
```
from ultralytics.yolo.data.annotator import auto_annotate

auto_annotate(data="ultralytics/assets", output_dir="../labels")
```
